test/elixir/test/auth_cache_test.exs

```diff
@@ -0,0 +1,215 @@
+      %{
+        :section => "chttpd_auth",
+        :key => "authentication_db",
+        :value => db_name
+      },
+      %{
+        :section => "chttpd_auth",
+        :key => "auth_cache_size",
+        :value => "3"
+      },
+      %{
+        :section => "httpd",
+        :key => "authentication_handlers",
+        :value => "{couch_httpd_auth, default_authentication_handler}"
+      },
+      %{
+        :section => "chttpd_auth",
+        :key => "secret",
+        :value => generate_secret(64)
+      }
+    ]
+
+    run_on_modified_server(server_config, fn -> test_fun(db_name) end)
+  end
+
+  defp generate_secret(length) do
+    tab = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/"
+
+    for _i <- 1..length, into: "" do
+      String.at(tab, trunc(Float.floor(:rand.uniform() * 64)))
+    end
+  end
+
+  defp hits() do
+    hits = request_stats(["couchdb", "auth_cache_hits"], true)
+    hits["value"] || 0
+  end
+
+  defp misses() do
+    misses = request_stats(["couchdb", "auth_cache_misses"], true)
+    misses["value"] || 0
+  end
+
+  defp logout(session) do
+    assert Couch.Session.logout(session).body["ok"]
+  end
+
+  defp login_fail(user, password) do
+    resp = Couch.login(user, password, :fail)
+    assert resp.error, "Login error is expected."
+  end
+
+  defp login(user, password) do
+    sess = Couch.login(user, password)
+    assert sess.cookie, "Login correct is expected"
+    sess
+  end
+
+  defp wait_until_compact_complete(db_name) do
+    info = info(db_name)
+
+    if info["compact_running"] do
+      wait_until_compact_complete(db_name)
+    end
+  end
+
+  defp assert_cache(event, user, password, expect \\ :expect_login_success) do
+    hits_before = hits()
+    misses_before = misses()
+
+    session =
+      case expect do
+        :expect_login_success -> login(user, password)
+        :expect_login_fail -> login_fail(user, password)
+        _ -> assert false
+      end
+
+    hits_after = hits()
+    misses_after = misses()
+
+    if expect == :expect_success do
+      logout(session)
+    end
+
+    case event do
+      :expect_miss ->
+        assert misses_after == misses_before + 1,
+               "Cache miss is expected for #{user} after login"
+
+        assert hits_after == hits_before,
+               "No cache hit is expected for #{user} after login"
+
+      :expect_hit ->
+        assert misses_after == misses_before,
+               "No cache miss is expected for #{user} after login"
+
+        assert hits_after == hits_before + 1,
+               "Cache hit is expected for #{user} after login"
+
+      _ ->
+        assert false
+    end
+  end
+
+  defp compact(db_name) do
+    resp = Couch.post("/#{db_name}/_compact")
+    assert resp.status_code == 202
+    resp.body
+  end
+
+  def save_doc(db_name, body) do
+    resp = Couch.put("/#{db_name}/#{body["_id"]}", body: body)
+    assert resp.status_code in [201, 202]
+    assert resp.body["ok"]
+    {:ok, resp}
```

Comment:
Did you consider returning the document itself instead of `{:ok, resp}`? It might simplify things a bit in the `test_fun`.
```
def save_doc(db_name, body) do
    resp = Couch.put("/#{db_name}/#{body["_id"]}", body: body)
    assert resp.status_code in [201, 202]
    assert resp.body["ok"]
    Map.put(body, "_rev", resp.body["rev"])
end
```